### PR TITLE
Add Use Cases for trips stuff

### PR DESCRIPTION
**Pull request by dellisd**
This restructures the Trips feature to use the UseCase pattern for passing data between the repository and the viewmodel. On top of this, it also now uses `Flow` from Kotlin's coroutines library to allow the viewmodel to react to updates to the cached trip data in the repository.

This also updates the Kotlin Coroutines library to `1.3.0-M1`.

# Checklist
- [x] I ran `./gradlew test connectedAndroidTest detekt`
- [x] I rebased off of `upstream/master`
- [ ] I added tests if applicable

### CODE REVIEW

**Comment by GustavoSanMartin** [comment]
I know it's already large, but you should really add tests in this pr!

File: app/src/main/java/ca/llamabagel/transpo/ui/trips/TripsViewModel.kt [hunk 1]
```diff
@@ -81,19 +87,27 @@ class TripsViewModel @Inject constructor(private val tripsRepository: TripsRepos
...
-        if (selectedRoutes.isEmpty()) return
+        /*if (selectedRoutes.isEmpty()) return
 
         val routeData = (apiData as Result.Success).data.routes
         _viewerData.value = routeData
             .filter { route -> selectedRoutes.contains(RouteSelection(route.number, route.directionId)) }
             .flatMap { route -> route.trips.map { trip -> TripUiModel(route, trip) } }
             .sortedBy { model -> model.trip.adjustedScheduleTime }
-            .map(::TripItem)
+            .map(::TripItem)*/
```
Comment (GustavoSanMartin) [inline app/src/main/java/ca/llamabagel/transpo/ui/trips/TripsViewModel.kt]: Are we still keeping this comment?

File: app/src/main/java/ca/llamabagel/transpo/ui/trips/adapter/TripsAdapter.kt [hunk 2]
```diff
@@ -33,10 +33,9 @@ class TripsAdapter(
         is TripItem -> R.layout.trip
     }
 
-
     override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
         when (holder) {
-            is SingleTripViewHolder -> holder.bind((getItem(position) as TripItem).tripUiModel)
+            is SingleTripViewHolder -> holder.bind((getItem(position) as TripItem))
```
Comment (GustavoSanMartin) [inline app/src/main/java/ca/llamabagel/transpo/ui/trips/adapter/TripsAdapter.kt]: I don't think we need two brackets here

File: app/src/main/java/ca/llamabagel/transpo/ui/trips/GetNextBusTripsUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        repository.getResultCache(stopId)
+            .asFlow()
+            .flowOn(Dispatchers.Default)
+            .map { response -> transformResponse(response) }
+            .flowOn(Dispatchers.Main)
+
+    private fun transformResponse(response: ApiResponse): List<TripAdapterItem> =
+            response.routes
+                .flatMap { route ->
+                    route.trips.map { trip -> TripItem(route, trip) }
```
Comment (GustavoSanMartin) [inline app/src/main/java/ca/llamabagel/transpo/ui/trips/GetNextBusTripsUseCase.kt]: Not a fan of this nested map but I don't see another way 😕

File: app/src/main/java/ca/llamabagel/transpo/data/TripsRepository.kt [hunk 4]
```diff
@@ -4,29 +4,90 @@
...
+    }
+
+    suspend fun getGroupByRoute(): Boolean =
+        withContext(Dispatchers.IO) { sharedPreferences.getBoolean(KEY_GROUP_BY_ROUTE, false) }
+
+    /**
+     * Rebroadcasts the last response
+     */
+    private fun rebroadcast() {
+        cachedResults.forEach { (_, v) ->
```
Comment (GustavoSanMartin) [inline app/src/main/java/ca/llamabagel/transpo/data/TripsRepository.kt]: I think it's a bit more readable if we only have 1 parameter in the `forEach`. Also maybe change `v` to something more descriptive?
```suggestion
        cachedResults.map { it.value }.forEach { v -> v.offer(v.value) }
```

File: app/src/main/java/ca/llamabagel/transpo/ui/trips/TripsViewModel.kt (hunk 1 above)
Comment (dellisd) [reply app/src/main/java/ca/llamabagel/transpo/ui/trips/TripsViewModel.kt]: No. it's been moved to a usecase, will remove

File: app/src/main/java/ca/llamabagel/transpo/data/TripsRepository.kt (hunk 4 above)
Comment (dellisd) [reply app/src/main/java/ca/llamabagel/transpo/data/TripsRepository.kt]: `.map { it.value }` is just the verbose version of `(_, v)`. Not very Kotlin-like. Plus doing a map creates a wasteful temporary object. I can give it a better name though.

Comment (dellisd) [reply app/src/main/java/ca/llamabagel/transpo/data/TripsRepository.kt]: I would just use `.forEachValue` but that requires Java 8 (i.e. SDK 24+)

**Pull request merged**